solr/core/src/java/org/apache/solr/cluster/placement/PropertyKeyFactory.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.cluster.placement;
+
+/**
+ * Factory used by the plugin to create property keys to request property values from Solr.<p>
+ *
+ * Building of a {@link PropertyKey} requires specifying the target (context) from which the value of that key should be
+ * obtained. This is done by specifying the appropriate {@link PropertyValueSource}.<br>
+ * For clarity, when only a single type of target is acceptable, the corresponding subtype of {@link PropertyValueSource} is used instead
+ * (for example {@link Node}).
+ */
+public interface PropertyKeyFactory {
+  /**
+   * Returns a property key to request the number of cores on a {@link Node}.
+   */
+  PropertyKey createCoreCountKey(Node node);
```

Comment:
IMHO this model is very limiting .. because every time we add another type of `PropertyKey` we have to change this interface.

How about something like this?

`
<T> T createPropertyKey(Node node, Class<T extends PropertyKey> propertyClass, String... props)
`

(or something like that ... not sure about the generics syntax here)

Reply:
If we add new types of `PropertyKeys` we will have to add implementations for these new keys. Wouldn't we need to touch the Solr codebase anyway? Clients (plugins) using the interface would have to know about the new implementation classes and update their code to use them. Technically they could pass a class name through config or other means to use new implementations without code change, but is it a realistic scenario? What would they do with these keys? What are the values these keys will fetch and how will they be used?
I'm not against making generic and highly flexible code but only if it's really needed. So if you have a real use case in mind that we should support, I'm open. Otherwise I'd rather keep things strongly typed for now (and as long as we only add stuff to these interfaces we're not breaking anything so we can add later).

Reply:
Ok.

Reply:
I think this part of the API is the most controversial, as it indeed leads to a large number of interfaces and impl. classes.

On one hand this indeed provides the super-strong typing, the burden is on Solr-core to provide the implementation of these interfaces so it doesn't complicate the lives of plugin devs. But on the other hand this may be taking things too far ...

I don't see yet how to simplify it without compromising the strong-typing too much - maybe just use a single `PropertyKey` interface + enum for key types and use one method like `PropertyKey createPropertyKey(PropertyKey.Type type, Node node, Object... params)`? This of course compromises the strong-typing, as it uses weakly-typed params and requires callers to know what parameters to supply ... so I'm not sure. But it would reduce the API surface significantly so maybe it's a compromise worth taking.

Reply:
Keep in mind it's not subclassing just for the pleasure of subclassing. In this example of `createCoreCountKey()`, we end up building an `AbstractNodePropertyKey.CoreCountImpl` and that's where the definition is that this key will correspond to `ImplicitSnitch.CORES`. If we make creation code generic, we'll have to move the actual differences in implementation to somewhere else.

Another key create by `createMetricKey` does even more in `AbstractNodePropertyKey.NodeMetricImpl` since it translates the abstracted registry (from `NodeMetricRegistry`) into the actual registry and assembles the metric name.

In other words, there's specific code to write regardless for each new key type assuming we don't want to expose the internal implementation of Solr (and parameters such as `"metrics:"` and `"totaldisk"` etc used in messages).